modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/consensus/ConsensusStateUpdater.scala

```diff
@@ -0,0 +1,182 @@
+
+import org.tessellation.ext.crypto._
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.node.NodeState
+import org.tessellation.schema.peer.PeerId
+import org.tessellation.sdk.domain.cluster.storage.ClusterStorage
+import org.tessellation.sdk.domain.consensus.ConsensusFunctions
+import org.tessellation.sdk.domain.gossip.Gossip
+import org.tessellation.sdk.infrastructure.consensus.message._
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
+import org.tessellation.security.signature.signature.{Signature, SignatureProof}
+
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+trait ConsensusStateUpdater[F[_], Key, Artifact] {
+
+  def tryAdvanceConsensus(
+    key: Key,
+    resources: ConsensusResources[Artifact]
+  ): F[Option[ConsensusState[Key, Artifact]]]
+
+  def tryFacilitateConsensus(
+    key: Key,
+    lastKeyAndArtifact: (Key, Artifact)
+  ): F[Option[ConsensusState[Key, Artifact]]]
+}
+
+object ConsensusStateUpdater {
+
+  def make[F[_]: Async: KryoSerializer: SecurityProvider, Event, Key: Show: Order: Next: TypeTag: ClassTag, Artifact <: AnyRef: Show: TypeTag](
+    consensusFns: ConsensusFunctions[F, Event, Key, Artifact],
+    consensusStorage: ConsensusStorage[F, Event, Key, Artifact],
+    clusterStorage: ClusterStorage[F],
+    gossip: Gossip[F],
+    keyPair: KeyPair,
+    selfId: PeerId
+  ): ConsensusStateUpdater[F, Key, Artifact] = new ConsensusStateUpdater[F, Key, Artifact] {
+
+    private val logger = Slf4jLogger.getLoggerFromClass(ConsensusStateUpdater.getClass)
+
+    def tryFacilitateConsensus(
+      key: Key,
+      lastKeyAndArtifact: (Key, Artifact)
+    ): F[Option[ConsensusState[Key, Artifact]]] =
+      consensusStorage
+        .updateState(key)(logIfUpdated(internalTryFacilitateConsensus(key, lastKeyAndArtifact)))
+
+    def tryAdvanceConsensus(
+      key: Key,
+      resources: ConsensusResources[Artifact]
+    ): F[Option[ConsensusState[Key, Artifact]]] =
+      consensusStorage
+        .updateState(key)(logIfUpdated(internalTryAdvanceConsensus(key, resources)))
+
+    private def logIfUpdated(fn: consensusStorage.StateUpdateFn): consensusStorage.StateUpdateFn =
+      fn(_, _).flatTap {
+        case Some((true, newState)) =>
+          logger.info(s"Consensus status for key ${newState.key.show} transitioned to ${newState.status.show}")
+        case _ => Applicative[F].unit
+      }
+
+    private def internalTryFacilitateConsensus(key: Key, lastKeyAndArtifact: (Key, Artifact))(
+      maybeState: Option[ConsensusState[Key, Artifact]],
+      setter: Option[ConsensusState[Key, Artifact]] => F[Boolean]
+    ): F[Option[(Boolean, ConsensusState[Key, Artifact])]] =
+      maybeState match {
+        case None =>
+          for {
+            peers <- clusterStorage.getPeers
+            facilitators = selfId :: peers.filter(_.state == NodeState.Ready).map(_.id).toList
```

Comment:
filtering should be made like `NodeState.ready(peers)` (see: `node.scala`).
We had issues with such explicit filtering peers' states on the current mainnet, when we were adding new states (deriving from Ready), filters were not working.

Reply:
Ok, I wasn't sure about this filtering either. Fixing it.